test/typecheck

```diff
@@ -0,0 +1,92 @@
+#! /usr/bin/python3
+
+import re
+import subprocess
+import sys
+
+# Run tsc as a type checker and throw away most of its output, in a
+# controlled way.
+
+# These errors are ignored for "relaxed" files. This is supposed to be
+# roughly equivalent to "noImplicitAny = false".
+
+ignored_codes = [
+    "TS2683",   # 'this' implicitly has type 'any' because it does not have a type annotation.
+    "TS7005",   # Variable '*' implicitly has an 'any[]' type.
+    "TS7006",   # Parameter '*' implicitly has an 'any' type.
+    "TS7008",   # Member '*' implicitly has an 'any[]' type.
+    "TS7009",   # 'new' expression, whose target lacks a construct signature, implicitly has an 'any' type.
+    "TS7010",   # '*', which lacks return-type annotation, implicitly has an 'any' return type.
+    "TS7015",   # Element implicitly has an 'any' type because index expression is not of type '*'.
+    "TS7016",   # Could not find a declaration file for module '*'...
+    "TS7019",   # Rest parameter '*' implicitly has an 'any[]' type
+    "TS7022",   # '*' implicitly has type 'any'...
+    "TS7023",   # '*' implicitly has return type 'any' because ...
+    "TS7024",   # Function implicitly has return type 'any' because ...
+    "TS7031",   # Binding element '*' implicitly has an 'any' type.
+    "TS7034",   # Variable '*' implicitly has type 'any' in some locations where its type cannot be determined.
+    "TS7053",   # Element implicitly has an 'any' type because expression of type 'any' can't be used to
+                # index type '*'.
+]
+
+
+def should_ignore(path):
+    if path.startswith("node_modules/"):
+        return True
+    return False
+
+
+is_relaxed_memo: dict[str, bool] = {}
+
+
+def is_relaxed(path):
+    if path in is_relaxed_memo:
+        return is_relaxed_memo[path]
+    with open(path) as fp:
+        n = 0
+        relaxed = False
+        while n < 50:
+            line = fp.readline()
+            if line is None:
+                break
+            if "@cockpit-ts-relaxed" in line:
+                relaxed = True
+                break
+            n += 1
+        is_relaxed_memo[path] = relaxed
+        return relaxed
+
+
+num_errors = 0
+
+
+def show_error(lines):
+    global num_errors
+    num_errors += 1
+    for line in lines:
+        sys.stdout.write(line)
+
+
+def consider(lines):
+    m = re.match("([^:]*)\\(.*\\): error ([^:]*): .*", lines[0])
+    if m and not should_ignore(m[1]):
+        relaxed = is_relaxed(m[1])
+        if not relaxed or m[2] not in ignored_codes:
+            show_error(lines)
```

Comment:
> Would it be possible to find the "relaxed" files in a first pass, and then call tsc --project tsconfig-relaxed.json on them instead? And call the strict ones with the other half? --help seems to suggest that this is possible.

I think it is hard to control the set of files that TSC produces errors for. The "includes" and "excludes" in the tsconfig.json file is only for the "root" files of your project, it will then follow "import"s and work on all those that it finds, regardless of whether they are in "includes" or even "excludes". The only real distinction is between *.ts and *.js files, and we can make it ignore all errors in *.js files via `checkJS: false`.

So, the machinery here for ignoring specific errors in specific files is something that we can't get from TSC otherwise.

And yes, running TSC twice on the same files with different strictness settings seems to be asking for trouble, since the inference rules seem to change.

I'd rather run TSC once on all files with the "best" strictness we want, and then decide independently which of those errors should prevent new code from entering "main". This also gives a stable situation for the LSP annotations in our editors. Both CI and LSP will use the exact same tsconfig.json and produce the same errors.